mia/medium/nintendo-64.cpp

```diff
@@ -4,8 +4,97 @@ struct Nintendo64 : Cartridge {
   auto load(string location) -> bool override;
   auto save(string location) -> bool override;
   auto analyze(vector<u8>& rom) -> string;
+  auto ipl2checksum(u32 seed, array_view<u8> rom) -> u64;
 };
 
+auto Nintendo64::ipl2checksum(u32 seed, array_view<u8> rom) -> u64
+{
+  auto rotl = [](u32 value, u32 shift) -> u32 {
+    return (value << shift) | (value >> (-shift&31));
+  };
+  auto rotr = [](u32 value, u32 shift) -> u32 {
+    return (value >> shift) | (value << (-shift&31));
+  };
+
+  auto csum = [](u32 a0, u32 a1, u32 a2) -> u32 {
+    if (a1 == 0) a1 = a2;
+    u64 prod = (u64)a0 * (u64)a1;
+    u32 hi = (u32)(prod >> 32);
+    u32 lo = (u32)prod;
+    u32 diff = hi - lo;
+    return diff ? diff : a0;
+  };
+
+  auto read = [&](u32 offset) -> u32 {
+      u32 data;
+      data  = rom[offset + 0] << 24;
+      data |= rom[offset + 1] << 16;
+      data |= rom[offset + 2] <<  8;
+      data |= rom[offset + 3] <<  0;
+      return data;
+  };
+
+  // create the initialization data
+  u32 init = 0x6c078965 * (seed & 0xff) + 1;
+  u32 data = read(0);
+  init ^= data;
+
+  // copy to the state
+  u32 state[16];
+  for (auto &s : state) s = init;
+
+  u32 dataNext, dataLast;
+  u32 dataIndex = 0;
+  u32 loop = 0;
+  while (1) {
+      loop++;
+      dataLast = data;
+      data = read(dataIndex);
```

Comment:
FYI, array_view has a built-in helper for big endian reads.
```
      data = rom.readm(4);
      rom += 4;
```

Reply:
Done, though `readm` already advances the pointer